model/convert/service_event.go

```diff
@@ -49,58 +49,159 @@ func convertServiceEventEpochSetup(event flow.Event) (*flow.ServiceEvent, error)
+	var dkgPhase3FinalView cadence.UInt64
+	var cdcClusters cadence.Array
+	var cdcParticipants cadence.Array
+	var foundFieldCount int
+
+	evt := cdcEvent.Type().(*cadence.EventType)
+
+	for i, f := range evt.Fields {
+		switch f.Identifier {
+		case "counter":
+			foundFieldCount++
+			counter, ok = cdcEvent.Fields[i].(cadence.UInt64)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"counter",
+					cdcEvent.Fields[i],
+					cadence.UInt64(0),
+				)
+			}
+
+		case "nodeInfo":
+			foundFieldCount++
+			cdcParticipants, ok = cdcEvent.Fields[i].(cadence.Array)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"participants",
+					cdcEvent.Fields[i],
+					cadence.Array{},
+				)
+			}
+
+		case "firstView":
+			foundFieldCount++
+			firstView, ok = cdcEvent.Fields[i].(cadence.UInt64)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"firstView",
+					cdcEvent.Fields[i],
+					cadence.UInt64(0),
+				)
+			}
+
+		case "finalView":
+			foundFieldCount++
+			finalView, ok = cdcEvent.Fields[i].(cadence.UInt64)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"finalView",
+					cdcEvent.Fields[i],
+					cadence.UInt64(0),
+				)
+			}
+
+		case "collectorClusters":
+			foundFieldCount++
+			cdcClusters, ok = cdcEvent.Fields[i].(cadence.Array)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"clusters",
+					cdcEvent.Fields[i],
+					cadence.Array{},
+				)
+			}
+
+		case "randomSource":
+			foundFieldCount++
+			randomSrcHex, ok = cdcEvent.Fields[i].(cadence.String)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"randomSource",
+					cdcEvent.Fields[i],
+					cadence.String(""),
+				)
+			}
+
+		case "DKGPhase1FinalView":
+			foundFieldCount++
+			dkgPhase1FinalView, ok = cdcEvent.Fields[i].(cadence.UInt64)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"dkgPhase1FinalView",
+					cdcEvent.Fields[i],
+					cadence.UInt64(0),
+				)
+			}
+
+		case "DKGPhase2FinalView":
+			foundFieldCount++
+			dkgPhase2FinalView, ok = cdcEvent.Fields[i].(cadence.UInt64)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"dkgPhase2FinalView",
+					cdcEvent.Fields[i],
+					cadence.UInt64(0),
+				)
+			}
+
+		case "DKGPhase3FinalView":
+			foundFieldCount++
+			dkgPhase3FinalView, ok = cdcEvent.Fields[i].(cadence.UInt64)
+			if !ok {
+				return nil, invalidCadenceTypeError(
+					"dkgPhase3FinalView",
+					cdcEvent.Fields[i],
+					cadence.UInt64(0),
+				)
+			}
+		}
 	}
-	setup.FirstView = uint64(firstView)
-	finalView, ok := cdcEvent.Fields[3].(cadence.UInt64)
-	if !ok {
-		return nil, invalidCadenceTypeError(
-			"finalView",
-			cdcEvent.Fields[3],
-			cadence.UInt64(0),
+
+	if foundFieldCount != expectedFieldCount {
+		return nil, fmt.Errorf(
+			"EpochSetup event required fields not found (%d != %d)",
+			foundFieldCount,
+			expectedFieldCount,
 		)
 	}
+
+	setup := new(flow.EpochSetup)
+
+	setup.Counter = uint64(counter)
```

Comment:
maybe skip the counter/firstView/finalView temp variable, and just directly assign the values into the setup struct

Reply:
Maybe not in this case because we need to first test if counter/firstView/finalView are of specific Cadence types and then cast them as Go types.